rfcs/websockets_py3.md

```diff
@@ -0,0 +1,45 @@
+# RFC 70: Add Python3 WebSockets dependency for WebDriver Bidi tests
+
+## Summary
+
+Add Py3-only [WebSockets client](https://github.com/aaugustin/websockets) that uses asyncio for WebDriver Bidi tests. 
+
+## Details
+
+### Background
+
+[WebDriver Bidi](https://w3c.github.io/webdriver-bidi/) builds and extends on [WebDriver](https://w3c.github.io/webdriver/).
+WebDriver Bidi uses WebSockets for bidirectional communication between remote end and local end.
+Existing pywebsocket3 library only serves as a WebSockets server. Therefore, a new WebSockets client dependency is needed in order to connect to WebDriver for Bidi communication.  
+
+### Proposal
+
+Add Py3-only WebSockets library linked in Summary.
+
+Add [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio) for convenient test marker.
+Note pytest-asyncio depends on later versions of pytest and pluggy so they also need to upgrade to compatible versions.
+
+### Example usage
+
+The [patch](https://github.com/web-platform-tests/wpt/pull/26510) 
+1. Adds a simple test to verify the new capability that would enable websocket connection for Bidi communication.
+1. Upgrades existing tools/third_party dependencies which are depended by the pytest-asyncio which provides convenient markers such as
+"@pytest.mark.asyncio" for writing tests.
+1. Makes use of the async/await syntax.
+1. Adds a BidiSession class in tools/webdriver/webdriver/client.py to be used
+in test fixture.
+1. Adds tests to verify the behavior of BidiSession across tests.
+
+A couple issues not solved in the patch:
+
+1. When upgrading pytest to 6.1.1 in example code, there was an error from tools/third_party/pytest/src/_pytest/assertion/rewrite.py
+about missing _pytest._version module.
+I commented out the import and hardcoded the version to 6.1.1.
+1. I installed the WebSockets and pytest-asyncio dependencies locally instead of including them in tools/third_party library.
+
+## Risks
+
+The current plan for python migration(I believe) is Py3-first by the end of the year and Py3-only Feb 2021.
+The patch is Py3-only and there is a gap of 3 months from now and then.
+Do we want to wait until then to add the dependencies?
+Or do we want to be able to exclude these dependencies for Py2?
```

Comment:
The February 2021 date is correct. Reference: the Py3 overview RFC, https://github.com/web-platform-tests/rfcs/pull/69/files

So it really comes down to how quickly we want/need webdriver bidi testing support, which I think is really something for webdriver folks to answer :). Internally to WPT, we likely won't be developing a webdriver bidi executor until late 2021 I expect (once bidi has standardized better).

Who are the correct decision makers for whether webdriver bidi testing support is needed before Feb 2021? @jgraham ?

Reply:
I think we will want to land this support before feburary, but I also think that we can make it opt-in (c.f. the quic support), or only enabled for the correct Python version. In practice it's expected that vendors and GH infra will be running Py3 by EoY so the gap is more like a few weeks.

Reply:
I talked to @Hexcles offline about "quick support" and "enabling for Py3".
I believe we can conditionally import "async" using six.PY3 flag and skip WebDriver tests that use async.
However, It's tricky to handle the third party dependency upgrade for pytest as the upgraded version is Py3 only.
I might just wait for "a few weeks" for the gap to fill.